Dining/views.py

```diff
@@ -185,19 +176,48 @@ def dispatch(self, request, *args, **kwargs):
         self.init_date()
         available_slots = DiningList.objects.available_slots(self.date)
         if available_slots <= 0:
-            return HttpResponseForbidden('No available slots')
+            error = _("No free slots availlable")
+            messages.add_message(request, messages.ERROR, error)
+            return HttpResponseRedirect(self.reverse('day_view', kwargs={}))
+
         if len(DiningList.objects.filter(date=self.date).filter(claimed_by=self.request.user)) > 0:
-            return HttpResponseForbidden('You already have dining slot claimed today')
+            error = _("You already have dining slot claimed today")
+            messages.add_message(request, messages.ERROR, error)
+            return HttpResponseRedirect(self.reverse('day_view', kwargs={}))
         return super().dispatch(request, *args, **kwargs)
 
 
 class EntryAddView(LoginRequiredMixin, DiningListMixin, TemplateView):
     template_name = "dining_lists/dining_entry_add.html"
     add_external_button_name = "addExternalButton"
 
+    def check_user_permission(self, request, user):
+        # If user is dining list owner or purchaser
+        if user == self.dining_list.claimed_by or user == self.dining_list.purchaser:
+            return True
+
+        if not self.dining_list.is_open():
+            # Dining list is closed. Go back to the info screen
+            error = _("Dining list is closed, people can no longer join")
+            messages.add_message(request, messages.ERROR, error)
+            return False
+
+        if not self.dining_list.has_room():
+            # dining list is full. Go back to the info screen
+            error = _("Dining list is already full, ask the chef personally if you want to join")
+            messages.add_message(request, messages.ERROR, error)
+            return False
+
+        # No problems, use can add people
+        return True
+
     def get(self, request, *args, **kwargs):
         context = self.get_context_data()
 
+        # Check permissions, if user has no access to this page, reject it.
+        if not self.check_user_permission(request, request.user):
```

Comment:
I am not sure how to do it in Django, but would it be possible to invert these access control settings? Right now we accept requests by default and manually check if the user should be rejected, so the access control settings are similar to 'Accept by default, reject for certain paths'. A safer setting would be 'Reject by default, accept for certain paths'. Is this possible in Django?

Reply:
It could be similar to the `LoginRequiredMixin` and `DiningListMixin`.

Reply:
I could write it in a mixin, but these specific checks aren't used in any other class and I see this not being used any time soon either. So placing it in a different class didn't make much sens to me tbh.
That said, I guess I could adjust it to do the checkups from the dispatch instead to apply on both 'get' and 'post'. Should I do this instead?

Reply:
Using `dispatch` similar to `LoginRequiredMixin` is better so that all HTTP methods are covered. You will however need to write a separate Mixin class in this case because of the method execution ordering. Because if you add the dispatch method in this class, if you call `super().dispatch()` at the start of the method, the whole get/post riddle will be executed before the check. On the other hand, if you first check the permissions before calling the super method, the Login and Dining mixins have not yet been processed, since that also happens in the `dispatch` method and is only executed when calling the super method. The solution that comes to mind therefore is to have a separate mixin class which implements `dispatch` (does not need to extend any class) and put it after the `DiningListMixin`.

Other note: the whole `get` method can be rewritten to `get_context_data`, just like e.g. `DayView`. That is a standard Django idiom which ensures that the same template context is available for every HTTP method, get and post in this case. (Again some literature :smile: : https://stackoverflow.com/a/36950584/2373688)

Reply:
I understand your argument on using mixin and while it is slightly cleaner to use it in a dispatch it doesn't make sense (at least to me) to have a part of the current EntryAddView class to be placed in a different class even though that content is only applicable to this very class and this class only.

I did implement get_context_data()